src/Build/Collections/PropertyDictionary.cs

```diff
@@ -113,10 +119,16 @@ internal PropertyDictionary(int capacity, IEnumerable<T> elements)
         {
             get
             {
-                lock (_properties)
+                _lock.EnterReadLock();
```

Comment:
Maybe worth creating a struct that implements IDisposable so you can use a using statement instead of adding a try/finally with new indent.